cmakelib/bindings/stack.go

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright 2019 The Kythe Authors. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package bindings
+
+type varStack struct {
+	p *varStack
+	v map[string]string
+}
+
+// Push constructs and returns a new child stack.
+func (vs *varStack) Push() *varStack {
+	return &varStack{vs, make(map[string]string)}
+}
+
+// Set sets the value associated with the key in the current scope.
+func (vs *varStack) Set(key, value string) {
+	vs.v[key] = value
+}
+
+// Get returns the associated value in the nearest scope or the empty string.
+func (vs *varStack) Get(key string) string {
```

Comment:
I recommend making the signature of this include a flag so you can distinguish "set to empty" from "unset": 

```go
func (vs *varStack) Lookup(key string) (string, bool) {…
```

If you need a shortcut for cases where you don't care, maybe `LookupDefault(key, "")`?

Reply:
This is intended to mirror the semantics of CMake variable lookup which does not distinguish between the empty string and missing keys.

Reply:
Ah, ok. It'd probably be sufficient to include that in the doc string for the method then.

Reply:
I can add the additional functionality, but it generally shouldn't be needed.